bftengine/src/bftengine/SigManager.cpp

```diff
@@ -167,46 +167,35 @@ SigManager::SigManager(PrincipalId myId,
   ConcordAssert(verifiers_.size() >= publickeys.size());
 }
 
-SigManager::~SigManager() {
-  delete mySigner_;
-  set<RSAVerifier*> alreadyDeleted;
-  for (pair<PrincipalId, RSAVerifier*> v : verifiers_) {
-    if (alreadyDeleted.find(v.second) == alreadyDeleted.end()) {
-      delete v.second;
-      alreadyDeleted.insert(v.second);
-    }
-  }
-}
-
 uint16_t SigManager::getSigLength(PrincipalId pid) const {
   if (pid == myId_) {
     return (uint16_t)mySigner_->signatureLength();
   } else {
-    auto pos = verifiers_.find(pid);
-    if (pos == verifiers_.end()) {
+    std::shared_lock lock(mutex_);
+    if (auto pos = verifiers_.find(pid); pos != verifiers_.end()) {
```

Comment:
nit: this way is similar to a for loop initialization, and is new in c++17, right?

Also, a question - I'm not familiar  with the client key exchange, but in case you have an old request signed with the old key, we assume retransmissions will work? Is the client's private key exchanged exactly at the same time? I see that retransmission enforces a new creation (and signing) of a message (throws TimeoutException to the caller), so the old message is not kept.

Reply:
If a client has changed its key, it should sign with a new one